exampleCustomAudioDevice/src/main/java/com/twilio/video/examples/examplecustomaudiodevice/FileAndMicAudioDevice.kt

```diff
@@ -0,0 +1,360 @@
+package com.twilio.video.examples.examplecustomaudiodevice
+
+import android.content.Context
+import android.media.AudioManager
+import android.media.AudioRecord
+import android.media.AudioTrack
+import android.media.MediaRecorder
+import android.os.Build
+import android.os.Handler
+import android.os.HandlerThread
+import android.os.Process
+import android.util.Log
+import androidx.annotation.RequiresApi
+import com.twilio.video.AudioDevice
+import com.twilio.video.AudioDeviceContext
+import com.twilio.video.AudioFormat
+import tvi.webrtc.ThreadUtils
+import java.io.DataInputStream
+import java.io.IOException
+import java.io.InputStream
+import java.nio.ByteBuffer
+
+class FileAndMicAudioDevice(private val context: Context) : AudioDevice {
+    private var keepAliveRendererRunnable = true
+
+    // Average number of callbacks per second.
+    private val BUFFERS_PER_SECOND = 1000 / CALLBACK_BUFFER_SIZE_MS
+    private var fileWriteByteBuffer: ByteBuffer? = null
+    private var writeBufferSize = 0
+    private var inputStream: InputStream? = null
+    private var dataInputStream: DataInputStream? = null
+    private var audioRecord: AudioRecord? = null
+    private var micWriteBuffer: ByteBuffer? = null
+    private var readByteBuffer: ByteBuffer? = null
+    private var audioTrack: AudioTrack? = null
+
+    // Handlers and Threads
+    private var capturerHandler: Handler? = null
+    private var capturerThread: HandlerThread? = null
```

Comment:
It's usually a good idea to use null safe types in properties like what you have here. In some cases you can use a `lateinit var` instead if you are absolutely certain that the variable will be initialized before it will be accessed. This is useful so you don't have to use the null safe operator `?` every time you access the variable. For example if you changed this variable to a lateinit it would look like `private lateinit var capturerThread: HandlerThread`